src/diffusion_for_multi_scale_molecular_dynamics/generators/adaptative_corrector.py

```diff
@@ -0,0 +1,257 @@
+
+        idx = index_i - 1  # python starts indices at zero
+        t_i = self.noise.time[idx].to(composition_i.X)
+        sigma_i = self.noise.sigma[idx].to(composition_i.X)
+        q_matrices_i = self.noise.q_matrix[idx].to(composition_i.X)
+        q_bar_matrices_i = self.noise.q_bar_matrix[idx].to(composition_i.X)
+        q_bar_tm1_matrices_i = self.noise.q_bar_tm1_matrix[idx].to(composition_i.X)
+
+        model_predictions_i = self._get_model_predictions(
+            composition_i, t_i, sigma_i, unit_cell, cartesian_forces
+        )
+
+        # Even if the global flag 'one_atom_type_transition_per_step' is set to True, a single atomic transition
+        # cannot be used at the last time step because it is necessary for all atoms to be unmasked at the end
+        # of the trajectory. Here, we use 'first' and 'last' with respect to a denoising trajectory, where
+        # the "first" time step is at index_i = T and the "last" time step is index_i = 1.
+        this_is_last_time_step = idx == 0
+        one_atom_type_transition_per_step = (
+            self.one_atom_type_transition_per_step and not this_is_last_time_step
+        )
+
+        # atom types update
+        a_im1 = self._atom_types_update(
+            model_predictions_i.A,
+            composition_i.A,
+            q_matrices_i,
+            q_bar_matrices_i,
+            q_bar_tm1_matrices_i,
+            atom_type_greedy_sampling=self.atom_type_greedy_sampling,
+            one_atom_type_transition_per_step=one_atom_type_transition_per_step,
+        )
+
+        if this_is_last_time_step:
+            assert (a_im1 != self.masked_atom_type_index).all(), \
+                "There remains MASKED atoms at the last time step: review code, there must be a bug or invalid input."
+
+        # in the adaptative corrector approach, there is no predictor step applied on the X component
+        composition_im1 = AXL(
+            A=a_im1, X=composition_i.X, L=unit_cell
+        )  # TODO : Deal with L correctly
+
+        if self.record:
+            # TODO : Deal with L correctly
+            composition_i_for_recording = AXL(
+                A=composition_i.A, X=composition_i.X, L=unit_cell
+            )
+            # Keep the record on the CPU
+            entry = dict(time_step_index=index_i)
+            list_keys = ["composition_i", "composition_im1", "model_predictions_i"]
+            list_axl = [
+                composition_i_for_recording,
+                composition_im1,
+                model_predictions_i,
+            ]
+
+            for key, axl in zip(list_keys, list_axl):
+                record_axl = AXL(
+                    A=axl.A.detach().cpu(),
+                    X=axl.X.detach().cpu(),
+                    L=axl.L.detach().cpu(),
+                )
+                entry[key] = record_axl
+            self.sample_trajectory_recorder.record(key="predictor_step", entry=entry)
+
+        return composition_im1
+
+    def corrector_step(
```

Comment:
This method also seems like it is very close to `LangevinGenerator.corrector_step`.

I propose that you create a method `LangevinGenerator._get_corrector_step_epsilon`. In `LangevinGenerator`, it could just return the correct entry from the tabulated values, and you could overload this method here to compute epsilon in a more sophisticated way.

We could get rid of the array `sqrt_2_epsilon` in `LangevinDynamics`: surely it's not taking a sqrt that slows us down.
<img width="850" alt="Screenshot 2024-12-02 at 11 40 43 AM" src="https://github.com/user-attachments/assets/60778fb8-7880-4eab-ae12-6b81fc309d80">


Again, this would avoid a lot of duplication.

Reply:
changed. I did not delete sqrt_2_epsilon yet. It requires checking a few tests to make sure it is not called anywhere. We could either do it right now or leave it as a  TODO as part of a code cleanup effort.